Initial ServiceWorkerWindowClient in a Home Screen web app launched to handle notificationclick handler is inert for a short period

beidson opened the pull request:
#### 6833b7d7f7beadc95bb5cccba523332d630d7e42
<pre>
Initial ServiceWorkerWindowClient in a Home Screen web app launched to handle notificationclick handler is inert for a short period
<a href="https://bugs.webkit.org/show_bug.cgi?id=252544">https://bugs.webkit.org/show_bug.cgi?id=252544</a>
rdar://105684663

Reviewed by Youenn Fablet.

There is a period of time between when the Networking process is told about a new ScriptExecutionContext
(when the main resource load begins) and when that ScriptExecutionContext actually exists (when the page load commits)

During that timespan, a ServiceWorker can discover the new ServiceWorkerClient represented by this Context,
and try to postMessage to it, and the Networking process will try to forward that message along to the appropriate
WebContent process, but... it won&apos;t find its target.

So the message is dropped on the floor.

This probably was technically always theoretical if a ServiceWorker was handling the fetch event for the main resource
load for the DocumentLoader. But it was likely not discovered because in those cases, ServiceWorkers probably aren&apos;t
that interested in postMessage()&apos;ing to the client.

In NotificationClick handlers - especially without a Fetch handler - the vulnerable timespan increases, as does the
likelihood that the ServiceWorker wants to postMessage() to this new client.

This patch fixes the bug by:
1 - Changing postMessage() from Networking to WebContent processes to return whether or not the target was found
2 - If the target was not found, the Networking process remembers the failed postMessage for future use
3 - When ServiceWorkerContainer::startMessages() is called, it first fetches pending messages from the Networking process
    before firing off its locally deferred messages.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/SWClientConnection.cpp:
(WebCore::SWClientConnection::postMessageToServiceWorkerClient):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/ServiceWorkerClientPendingMessage.h: Added.
* Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
(WebCore::ServiceWorkerContainer::startMessages):
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::getServiceWorkerClientPendingMessages):
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::registerServiceWorkerClient):
(WebCore::SWServer::unregisterServiceWorkerClient):
(WebCore::SWServer::addServiceWorkerClientPendingMessage):
(WebCore::SWServer::releaseServiceWorkerClientPendingMessage):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::postMessageToServiceWorkerClient):
(WebKit::WebSWServerConnection::getServiceWorkerClientPendingMessages):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::getServiceWorkerClientPendingMessages):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:

Canonical link: <a href="https://commits.webkit.org/262711@main">https://commits.webkit.org/262711@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8cd522c941628beab074b4ea81dcc6fac7a97ba9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2142 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2175 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2241 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3065 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2190 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2116 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2255 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2220 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1943 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2161 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1917 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1916 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2920 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1897 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1801 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1974 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3085 "Passed tests") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1945 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1760 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1963 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1907 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/589 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1889 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2079 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
```diff
@@ -386,7 +386,14 @@ void WebSWServerConnection::postMessageToServiceWorkerClient(ScriptExecutionCont
+        if (result)
+            return;
+
+        server().addServiceWorkerClientPendingMessage(destinationContextIdentifier, { message, sourceServiceWorkerData, sourceOrigin });
```
I am not sure this ensures message ordering, say:
- a first message is sent, document is not there, async reply is false, first message is stored in network process
- a second message is sent, document is there this time, async reply is true, second message is stored in web process
- getServiceWorkerClientPendingMessages message is sent by web process
- second message gets fired before first message

One possible solution:
- When registering service worker clients at fetch event time, mark the service worker client as not yet ready to receive messages (in WebSWServerConnection::createFetchTask)
- When posting a message to a service worker client, check whether the client is ready. If not, store the message
- When client is created/ready to receive messages, it should already call registerServiceWorkerClientData, at which point we can push all the messages if any.

youennf replied:
Looking at service worker container code, I see this patch is preserving the ordering so that is good.
One potential risk though is if ServiceWorkerContainer is not created (or not created immediately while parsing the document).
In that case, the messages would remain in the network process until ServiceWorkerContainer is created.
It seems that, once the document is parsed, if ServiceWorkerContainer is not created, we should just drop the messages.
This might be observable behaviour.

beidson replied:
The potential mis-ordering you suggest is accounted for:
Because the network process messages are known to be stored before web process messages, and therefore explicitly fired before them.

(This occurs inside ServiceWorkerContainer::startMessages())

youennf commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -553,6 +553,11 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
 
+void WorkerSWClientConnection::getServiceWorkerClientPendingMessages(ScriptExecutionContextIdentifier, CompletionHandler<void(Vector<ServiceWorkerClientPendingMessage>&&)>&& completionHandler)
+{
+    completionHandler({ });
```
We probably have the same problem with workers and shared workers.

beidson replied:
I don't see how SharedWorkers can have this issue, as they cannot come into existence until *after* a document page load occurs, so their link to at least one parent ScriptExecutionContext is already properly established.

youennf replied:
We can probably post message to a worker or a shared worker as soon as the fetch event for their corresponding script is fired, ending up in the same issue as documents.
Worth adding a test and checking what other browsers are doing.

beidson replied:
While a ServiceWorker is handling a fetch event for a main resource, there cannot be a worker or (new) shared worker in existence before the DocumentLoader even commits. After the DL commits the time window for this bug is over.

(SharedWorkers that exist because of a different Document on the origin can be found, but also won't suffer this problem)

youennf replied:
It is worth testing to be sure, but the fetch event fired for a worker script fetch should expose the worker ID as `resultingClientId`. The worker is not yet created at this time since the script is not there but you can probably postMessage to it.

youennf commented on Source/WebCore/workers/service/server/SWServer.cpp:
```diff
@@ -1663,6 +1664,17 @@ void SWServer::fireFunctionalEvent(SWServerRegistration& registration, Completio
 
+void SWServer::addServiceWorkerClientPendingMessage(ScriptExecutionContextIdentifier contextIdentifier, ServiceWorkerClientPendingMessage&& message)
+{
+    auto addResult = m_clientPendingMessagesById.ensure(contextIdentifier, [] { return Vector<ServiceWorkerClientPendingMessage>(); });
```
Isn't there a chance that postMessage is called in service worker but client is already unregistered in networking process?
In that case, we might store the messages in m_clientPendingMessagesById forever.
We could have this Vector in a map that is populated when registering/unregistering clients.
If the map does not have the corresponding client, the message is dropped.
Or at least check that the service worker client is registered before appending the message, say in m_clientsById map. Or add an empty entry in m_clientPendingMessagesById as part of registerServiceWorkerClient and not call ensure in this method.

beidson replied:
Fair! Will totally update.

youennf reviewed (commented):
Can we add some tests?

youennf reviewed (approved):
This is ok as a temporary stopgap but I think we need to support workers as well and make sure to flush messages as soon as the document is ready (and not at the time service worker container is created).
A simpler approach that would solve both things is for all clients is to store a map<pending client ids, messages> in SWServer, and when each client is calling updateServiceWorkerClientData(), we send all buffered messages.
We probably also need to add a test. I'll add in another comment a test that I think should exercise that code path. Let me know whether it actually works, and if so, please include it in the landing patch.

beidson committed:
Initial ServiceWorkerWindowClient in a Home Screen web app launched to handle notificationclick handler is inert for a short period
https://bugs.webkit.org/show_bug.cgi?id=252544
rdar://105684663

Reviewed by Youenn Fablet.

There is a period of time between when the Networking process is told about a new ScriptExecutionContext
(when the main resource load begins) and when that ScriptExecutionContext actually exists (when the page load commits)

During that timespan, a ServiceWorker can discover the new ServiceWorkerClient represented by this Context,
and try to postMessage to it, and the Networking process will try to forward that message along to the appropriate
WebContent process, but... it won't find its target.

So the message is dropped on the floor.

This probably was technically always theoretical if a ServiceWorker was handling the fetch event for the main resource
load for the DocumentLoader. But it was likely not discovered because in those cases, ServiceWorkers probably aren't
that interested in postMessage()'ing to the client.

In NotificationClick handlers - especially without a Fetch handler - the vulnerable timespan increases, as does the
likelihood that the ServiceWorker wants to postMessage() to this new client.

This patch fixes the bug by:
1 - Changing postMessage() from Networking to WebContent processes to return whether or not the target was found
2 - If the target was not found, the Networking process remembers the failed postMessage for future use
3 - When ServiceWorkerContainer::startMessages() is called, it first fetches pending messages from the Networking process
    before firing off its locally deferred messages.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/SWClientConnection.cpp:
(WebCore::SWClientConnection::postMessageToServiceWorkerClient):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/ServiceWorkerClientPendingMessage.h: Added.
* Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
(WebCore::ServiceWorkerContainer::startMessages):
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::getServiceWorkerClientPendingMessages):
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::registerServiceWorkerClient):
(WebCore::SWServer::unregisterServiceWorkerClient):
(WebCore::SWServer::addServiceWorkerClientPendingMessage):
(WebCore::SWServer::releaseServiceWorkerClientPendingMessage):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::postMessageToServiceWorkerClient):
(WebKit::WebSWServerConnection::getServiceWorkerClientPendingMessages):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::getServiceWorkerClientPendingMessages):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:

Canonical link: https://commits.webkit.org/262711@main
